middleware/motore/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+
+/// The side where the middleware is deplyed.
+#[derive(Debug, Copy, Clone)]
+pub enum ServiceRole {
+    Server,
+    Client,
+}
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// It is used to extractor a resource name from requests for sentinel. If you the service request is [`http::Request`](https://docs.rs/http/latest/http/request/struct.Request.html),
+/// and you are using nightly toolchain, you don't need to provide a sentinel resource name extractor. The middleware will automatically extract the request uri.
+type Extractor<Cx, R> = fn(&Cx, &R) -> String;
+
+/// The fallback function when service is rejected by sentinel.
+type Fallback<S, Cx, R> =
+    fn(&Cx, &R, &sentinel_core::Error) -> Result<<S as Service<Cx, R>>::Response, BoxError>;
+
+/// The sentinel middleware service in motore.
+pub struct SentinelService<S, Cx, R>
+where
+    S: Service<Cx, R>,
+{
+    pub(crate) inner: S,
+    pub(crate) extractor: Option<Extractor<Cx, R>>,
+    pub(crate) fallback: Option<Fallback<S, Cx, R>>,
+    traffic_type: sentinel_core::base::TrafficType,
+}
+
+// seems rustc cannot derive `Clone` trait for us,
+// implement it by hands
+impl<S, Cx, R> Clone for SentinelService<S, Cx, R>
+where
+    S: Service<Cx, R> + Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            extractor: self.extractor.clone(),
+            fallback: self.fallback.clone(),
+            traffic_type: self.traffic_type,
+        }
+    }
+}
+
+impl<S, Cx, R> SentinelService<S, Cx, R>
+where
+    S: Service<Cx, R> + Clone,
+{
+    pub fn new(inner: S, role: ServiceRole) -> Self {
+        Self {
+            inner,
+            extractor: None,
+            fallback: None,
+            traffic_type: {
+                match role {
+                    ServiceRole::Server => sentinel_core::base::TrafficType::Inbound,
+                    ServiceRole::Client => sentinel_core::base::TrafficType::Outbound,
+                }
+            },
+        }
+    }
+
+    pub fn with_extractor(mut self, extractor: Extractor<Cx, R>) -> Self {
+        self.extractor = Some(extractor);
+        self
+    }
+
+    pub fn with_fallback(mut self, fallback: Fallback<S, Cx, R>) -> Self {
+        self.fallback = Some(fallback);
+        self
+    }
+}
+
+macro_rules! deal_with_sentinel {
+    ($resource:ident,$self:ident,$cx:ident,$req:ident) => {{
+        let entry_builder = EntryBuilder::new($resource).with_traffic_type($self.traffic_type);
+        match entry_builder.build() {
+            Ok(entry) => {
+                let fut = $self.inner.call($cx, $req).await.map_err(Into::into);
+                entry.exit();
+                fut
+            }
+            Err(err) => match $self.fallback {
+                Some(fallback) => fallback(&$cx, &$req, &err),
+                None => Err(Into::<BoxError>::into(err).into()),
+            },
+        }
+    }};
+}
+
+#[cfg(feature = "volo")]
```

Comment:
had tried features like `negative_impl`, but finally gave up and specialized the generics by hand, see this [blog](https://forsworns.github.io/zh/blogs/20221108/)